src/node/memory_leak/memory_leak_tests.js

```diff
@@ -48,42 +48,60 @@ const DATA_POINTS = 10;
  * This function attempts to determine approximately whether a given operation
  * leaks any memory. In theory, if a single fixed operation leaks memory, then
  * the increase in memory usage should be approximately proportional to the
- * number of times that operation is run. We estimate this by calling the
- * operation N times and recording the increase in memory usage (if any), then
- * calling it again another 2 * N times, again recording the increase in memory
- * usage relative to the starting value. To account for noise, we make a
- * conservative estimate and say that it leaked if the additional memory used
- * after a total of 3 * N calls is at least twice the additional memory used
- * after just N calls. We try to choose N that is large enough that its
- * operations are the dominant factor, and small enough that if there is a leak
- * the process will not run out of memory.
+ * number of times that operation is run. We estimate this by running the
+ * operation NUM_RUNS times and recording the memory used since the beginning
+ * of the entire test. Then we repeat that DATA_POINTS times. As a heuristic
+ * to check for linear growth, we check each data point against a linear
+ * extrapolation of the first data point to the previous index (e.g. we check
+ * the fifth data point against 4 times the first data point).
+ * We run NUM_RUNS functions before we start recording data as a warm up to
+ * avoid outliers caused by one-time initialization that the first calls may
+ * cause.
  * @param {function(function(Error))} checkFun The function to check for leaks
  * @param {function(Error)} cb The callback to pass the result to
  */
 function doesLeak(checkFun, cb) {
   function wrappedCheckFun(id, next) {
     checkFun(next);
   }
-  const startRss = process.memoryUsage().rss;
-  function runManyAndCheckUsage(id, next) {
+  function warmUp(run) {
     async.times(NUM_RUNS, wrappedCheckFun, function(err) {
```

Comment:
I think the actual warmup is done within this `async.times(NUM_RUNS...)`, and then the rest of warmUp id doing the test? Can we rename `warmUp` to something like `warmpAndThenTest`?